include/LightGBM/config.h

```diff
@@ -397,6 +397,9 @@ struct Config {
   // desc = LightGBM allows you to provide multiple evaluation metrics. Set this to ``true``, if you want to use only the first metric for early stopping
   bool first_metric_only = false;
 
+  // desc = LightGBM allows you to provide multiple evaluation metrics. Set this to a specific metric name, if you want to use only this metric for early stopping
+  std::string chosen_metric_early_stopping;
```

Comment:
Thinking about this some more... I don't think we should add this as a parameter in LightGBM's C API.

Right now, LightGBM (across all its interfaces), has this mix of behaviors:

* you can provide multiple metrics via the `metric` parameter
* if you set `early_stopping_rounds > 0` and provide any validation sets, LightGBM will try to perform early stopping based on **all metrics** and **all validation sets**
    * ... unless you set `first_metric_only = true`, in which case LightGBM will perform early stopping on only 1 metric (but still for all validation sets)

related: #6360 

Two types of behavior rely on that `metric` parameter:

* which metrics should be computed and logged/recorded during training?
* which metrics should be used for early stopping?

We still want to provide the ability to record multiple metrics during training.

In addition, the CLI and C API don't have a concept of "callbacks", so a parameter `metric_name` that only accepts a single metric wouldn't be sufficient for them if they want to perform early stopping on the basis of multiple metrics.

We also have to think carefully about what breaking changes (if any) to make to LightGBM's existing behavior of automatically performing early stopping on **all** metrics if you enable early stopping at all.

I'm not sure what direction to set you on... need to think about this for a few days. 

This is has been a quite complicated part of LightGBM's interface, I'd like to simplify it and give people finer control, but also do that in a way that minimizes the number of breaking changes made.

For example, maybe we could turn off the "automatically add the early stopping callback based on `params`" behavior if *any* `lgb.early_stopping` callbacks are passed through `callbacks`.

Reply:
Thanks for your explanations, I now also realize all the implications of this change, adjusting the python part with others! 

I also understand that being able to specify the `metric_name `in the parameters dict would be preferable, as other early stopping parameters can be specified here as well. However feel free to tell me to undo the changes outside of the Callback class, if it helps to split this in different PRs. 
My tests with the callback API changes alone have the expected behaviour.

I will try to take a bit more look at the C API and give you my 2cents during the week-end about how the change could be implemented. I don't expect to come up with the solution, but I guess it could help you to take a decision.

Reply:
Sure, thanks! I just want to be sure we're respectful of your time and limit how often we ask you to do something and then to undo it.

This is a part of LightGBM (and the Python package) that has to be handled with extreme care. Early stopping is a very important part of training GBDTs, and lots of existing code out in the world relies on the existing behavior.

If you want some more background on that,you might find this discussion useful: #5808